src/data/nat/factorial/cast.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import ring_theory.polynomial.pochhammer
+
+/-!
+# Cast of factorials
+
+This file allows calculating factorials (including ascending and descending ones) as elements of a
+semiring.
+-/
+
+open_locale nat
+
+variables (S : Type*)
+
+namespace nat
+
+section semiring
+variables [semiring S] (a b : ℕ)
+
+lemma cast_asc_factorial :
+  (a.asc_factorial b : S) = (pochhammer S b).eval (a + 1) :=
+by rw [←pochhammer_nat_eq_asc_factorial, pochhammer_eval_cast, nat.cast_add, nat.cast_one]
+
+lemma cast_desc_factorial :
+  (a.desc_factorial b : S) = (pochhammer S b).eval (a - (b - 1) : ℕ) :=
+begin
+  rw [←pochhammer_eval_cast, pochhammer_nat_eq_desc_factorial],
+  cases b,
+  { simp_rw desc_factorial_zero },
+  simp_rw [add_succ, succ_sub_one],
+  obtain h | h := le_total a b,
+  { rw [desc_factorial_of_lt (lt_succ_of_le h), desc_factorial_of_lt (lt_succ_of_le _)],
+    rw [nat.sub_eq_zero_of_le h, zero_add] },
+  { rw nat.sub_add_cancel h }
+end
+
+lemma cast_factorial :
+  (a! : S) = (pochhammer S a).eval 1 :=
+by rw [←zero_asc_factorial, cast_asc_factorial, cast_zero, zero_add]
+
+end semiring
+
+section ring
+variables [ring S] (a b : ℕ)
+
+lemma cast_desc_factorial_two :
```

Comment:
Shouldn't this just be a lemma about `a.desc_factorial : ℕ`? (Which can then be combined with `norm_cast`, or `cast_mul` lemmas.

Reply:
No, that's precisely why I need it. This is because the `-1` on the RHS is **not** truncated substraction anymore. And it still works because whenever `(a - 1 : ℕ) : S` is not equal to `a - 1 : S`, the other factor is `0` anyway.

That's also why I'm not providing the corresponding `nat.asc_factorial` lemma.

Reply:
Ok, thanks. That makes a lot of sense.
```suggestion
/- Convenience lemma. The `a - 1` is not using the truncated subtraction that it does use in `ℕ`. -/
lemma cast_desc_factorial_two :
```